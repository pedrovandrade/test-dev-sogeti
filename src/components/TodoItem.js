import React, { useState, useEffect } from 'react';
import { Link } from "react-router-dom";

export default function TodoItem(props) {
  /** ******** REACT HOOKS ********* */
  const [done, setDone] = useState(false);
  const [pageLoaded, setPageLoaded] = useState(false);

  function checkItem(event) {
    if (event.target.checked) {
      setDone(true);
    } else {
      setDone(false);
    }
  }

  /** ******** REACT EFFECTS ******** */
  // Page load
  useEffect(() => {
    setPageLoaded(true);
  }, []);

  // Throw the checked item at the bottom of the list if done or at the top if
  // it's still to do
  useEffect(() => {
    if (!pageLoaded) return; // To prevent useEffect calls from page load
    
    const index = props.index;
    const checkedElem = props.listData[index];
    if (done) {
      checkedElem.state = 'done';
      const newArray = props.listData.filter((elem) => {
        return elem !== props.listData[index];
      }).concat(checkedElem);
      props.setListData(newArray);
    } else {
      checkedElem.state = 'todo';
      const newArray = [checkedElem].concat(props.listData.filter((elem) => {
        return elem !== props.listData[index];
      }));
      props.setListData(newArray);
    }
  }, [done]);

  // Change the corresponding checkbox state when a list element is displaced
  useEffect(() => {
    const index = props.index;
    if (props.listData[index].state === 'done') {
      setDone(true);
    } else {
      setDone(false);
    }
  }, [props.listData]);

    return (
      <li>
        <label>
          <input type='checkbox' onChange={checkItem} checked={props.data.state === 'done'} />
          <Link to={`/details/${props.data.uuid}`} style={{ color: 'black', textDecoration: 'none' }}>
            <div className={`todo-item${props.data.state === 'done' ? ' checked' : ''}`}>{props.data.title}</div>
          </Link>
          <button className='delete-button' type='button' onClick={props.deleteItem} ></button>
        </label>
      </li>
    );
}
