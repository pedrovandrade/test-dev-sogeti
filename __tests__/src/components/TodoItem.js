import React from 'react';
import renderer from 'react-test-renderer';
import TodoItem from '../../../src/components/TodoItem';
import { MemoryRouter } from 'react-router-dom';

describe('src/components/TodoItem.js', () => {
  it('should expose a function component', () => {
    expect(TodoItem).toBeInstanceOf(Function);
  });
  it('should render correctly', () => {
    const mockList = [
      {
        title: 'Do groceries',
        state: 'todo',
        description: 'Go to the supermarket and buy eggs, butter, sugar, fruits, soap and toillete paper',
        uuid: '8348db18-83ba-11ec-a8a3-0242ac120002'
      },
      {
        title: 'Wash car',
        state: 'todo',
        description: 'Take the car to the gas station and leave it there for 30 minutes in the express washing',
        uuid: '9c377382-83ba-11ec-a8a3-0242ac120002'
      },
      {
        title: 'Clean house',
        state: 'todo',
        description: 'Pass vaccum, clean the surfaces, clean the toillete and pass water on the floor',
        uuid: 'a081b5ec-83ba-11ec-a8a3-0242ac120002'
      }
    ];

    const tree = renderer
    .create(
      <MemoryRouter>
        <TodoItem
          data={mockList[0]}
          deleteItem={() => null}
          listData={mockList}
          setListData={() => null}
          index={0}
          key={0}
        />
      </MemoryRouter>
    )
    .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
