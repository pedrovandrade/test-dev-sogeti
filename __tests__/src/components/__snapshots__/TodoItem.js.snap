// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/TodoItem.js should render correctly 1`] = `
<li>
  <label>
    <input
      checked={false}
      onChange={[Function]}
      type="checkbox"
    />
    <a
      href="/details/8348db18-83ba-11ec-a8a3-0242ac120002"
      onClick={[Function]}
      style={
        Object {
          "color": "black",
          "textDecoration": "none",
        }
      }
    >
      <div
        className="todo-item"
      >
        Do groceries
      </div>
    </a>
    <button
      className="delete-button"
      onClick={[Function]}
      type="button"
    />
  </label>
</li>
`;
